src/Xamarin.Android.Build.Tasks/Utilities/ELFHelper.cs

```diff
@@ -7,18 +8,28 @@
 using Microsoft.Android.Build.Tasks;
 using Microsoft.Build.Utilities;
 
+using ELFSymbolType = global::ELFSharp.ELF.Sections.SymbolType;
+
 namespace Xamarin.Android.Tasks
 {
 	static class ELFHelper
 	{
+		static ConcurrentDictionary<string, bool> checkedDSOs = new ConcurrentDictionary<string, bool> (StringComparer.OrdinalIgnoreCase);
```

Comment:
Do we *actually* want static mutable data?  We've previously had bugs due to static mutable data in this assembly: https://github.com/xamarin/xamarin-android/issues/5996

Reply:
`ConcurrentDictionary` is designed for this kind of sharing scenario, it should be fine.

Reply:
Respectfully disagree?  `ConcurrentDictionary` is designed to be thread-safe.  That doesn't describe or answer interactions with AppDomains and the IDE and user behavior and caching.

For example -- untested, but this "feels" right -- assume you have a Solution with two projects, `App.dll` and `Lib.dll`.

Initially, `Lib.dll` is *empty*: no classes, just e.g. `@(EmbeddedResource)` or whatever.

AOT is enabled.

At this point, AOT'ing `Lib.dll` *should* result in `Lib.dll` having *no* AOT members, as it's empty.

Developer updates `Lib.dll` so that it now contains types & methods.  Now, AOT'ing `Lib.dll` *should* result in a non-empty `libaot-Lib.dll.so` file, as it contains code.

How does `ELFHelper.checkedDSOs` get updated in this scenario?

If you have a separate process per build, it's moot: you have an empty `ELFHelper.checkedDSOs` dict at the beginning of every build.

*Within the IDE*, ***I don't know*** what will happen.  Historically (?), `Xamarin.Android.Build.Tasks.dll` is kept "alive" once loaded; *I imagine* that *within the IDE*, `ELFHelper.checkedDSOs` will remain loaded once set, *possibly across projects*.  (Again, no idea.  @jonathanpeppers? @dellis1972?)  The lack of knowledge is concerning, regardless.

Reply:
Hm, that scenario might be valid. I've never thought about the IDE in this context.

Reply:
I've decided to drop the caching, I like the simplicity of `ELFHelper.IsEmptyAOTLibrary` being static and the caching might not even make the builds that much faster (if at all, locking may make it potentially slower)